### ISSUE TITLE
Omission of popular contracts with long vanity addresses

### ISSUE THREAD

**Issue by perama-v** [TrueBlocks/trueblocks-core#3211]
## Problem

Here is an interesting address: `0x00000000000000ADc04C56Bf30aC9d3c0aAF14dC`

Technically omitted from the UnchainedIndex by definition (`v0.51.0-beta` starts with 14 zeros). 

The contract is Seaport 1.5, used by OpenSea as of May, has more than 1 million transactions. 
With this many transactions, it represents a significant omission to the index.

```console
Smallest permitted address
0x00000000000000ffffffffffffffffffffffffff
Seaport
0x00000000000000ADc04C56Bf30aC9d3c0aAF14dC
```
## Solution

One simple way could be to add an inclusion clause for deployed contracts. 

1. When chain is being chifra scraped it looks at contracts as they are deployed.
2. Detect any deployed contract that has an address that is invalid by the normal rules of the UnchainedIndex (such as too many leading zeros)
3. Add the address to "deployed vanity contracts" special list that is published alongside the unchained index, referenced in the manifest.
4. When looking for appearances, at the step when vanity detection is applied, permit if on the special list.

The main insights are: It costs money to deploy and money to grind a vanity address. So this is an anti-spam mechanism for this special list. The list is probably very small (?tens of addresses total)

The main benefit is that the index includes addresses such as the above seaport address.

## Alternative approaches

Some alternatives, each with their own problems
- Leave out of index and accept that some edge cases exist 
- Pick favourites 
- Rewrite old index chunks based on usage thresholds (such as a clause in the index definition to include an address with more than x transactions, even if not valid based on leading zeros).

**Comment by tjayrush** [comment]
First of all, thanks so much for this amazing issue.

Second, addressing each of your three options:

> Leave out of index and accept that some edge cases exist

We call these addresses (those below `0x00....ffff` as above) **baddresses** (bad addresses). It's a vanishingly small percentage of the total. I forget exactly, but it's easy to reproduce the number.) I never loved this as it's totally arbitrary, but without the idea of a `baddress` every 20-byte string is a valid address. We chose this number from two directions: (1) how high can we go without excluding too many addresses, and (2) how low can we go without including too many things that are clearly not addresses. (for example, 0x000....00001 -- boolean true -- is not an address). The choice we made was totally arbitrary and I'm totally open to discussing changes to that number.


> Pick favourites

This is a bad idea. Very little likelyhood of doing this in a way that changes the existing index...

However, I can see doing something like an "addendum" index where the end user chooses to re-scrape the entire chain looking for a list of specific `baddresses` (for example `chifra scrape --addendum <addresses>`.

This could create a different index than the existing one (say in a folder called `addendum` instead of `unchained`). It could be identically shaped and easily added given that it wouldn't be shared in the same way as `unchained`.

But `chifra export` could easily be modified to include an additional path to search. People could share these addendums, I suppose by publishing their own `addendum` databases.

> Rewrite old index chunks based on usage thresholds (such as a clause in the index definition to include an address with more than x transactions, even if not valid based on leading zeros).

Keeping track of how many is almost as onerous as including every `baddress` simply by default. I'm not sure that's desirable. For example, `0x0000....00001` boolean true appears billions of times, I'm sure. Probably blows out on a laptop. Also, all the precompiles are excluded.

**Comment by tjayrush** [comment]
Addressing your main suggestion:

> One simple way could be to add an inclusion clause for deployed contracts.

We already include all deployed addresses, I think. If not, this is a bug. When we scrape, we look at certain places in the data and call them `explicit` addresses. (For example, the `contractAddress` on the receipt. These `explicit` address do not go through the `baddress` identifier (or, at least, they shouldn't). We also look in the trace data's `action.address`.

> When chain is being chifra scraped it looks at contracts as they are deployed.

We should be doing this. If not, this is a bug.

> Detect any deployed contract that has an address that is invalid by the normal rules of the UnchainedIndex (such as too many leading zeros)

Yes.

> Add the address to "deployed vanity contracts" special list that is published alongside the unchained index, referenced in the manifest.

I call this an `adendum` above, but this shouldn't be needed as we should already be finding these deployed addresses through `action.address`.

> When looking for appearances, at the step when vanity detection is applied, permit if on the special list.

I'm not a big fan of 'special lists'. They are 100% impossible to maintain because there's always a new one and the old ones fade away. Any solution must be 100% fully automatic.

> The main insights are: It costs money to deploy and money to grind a vanity address. So this is an anti-spam mechanism for this special list. The list is probably very small (?tens of addresses total)

I'm not totally opposed to 10s of items. That can be a fairly simple configuration item, but that opens up that different people use a different list and then the idea that all indexes produced for `unchained` index are identical is under fire. Perhaps a better idea is a separate `adendum` index.

The main benefit is that the index includes addresses such as the above seaport address.

**Comment by tjayrush** [comment]
Would it be possible for you to find the exact deployment transaction for this particular address? That would let me debug it a bit and I could figure out why it's not already found if it appears anywhere....

Oh...wait...something just dawned on me...we might see the deployment, but all subsequent appearance would still be lost...but then...if we kept a list of deployed contracts so we didn't miss them...that would grow to millions and now we've killed running on a laptop in a reasonable way.

The more I think of it, the more I like the `adendum` index which sits beside the "real" index in a separate folder and doesn't really have any effect on the `unchained` part.

**Comment by perama-v** [comment]
> Would it be possible for you to find the exact deployment transaction for this particular address? 

0xa7f75a8b23c0f3150b7116aae930101f42996cb35e3e596a646156926350933a

**Comment by perama-v** [comment]
Ok I forgot that baddress definition only applies to implicit address appearances. So seaport would be included if appearing in explicit locations: 
- tx to
- internal tx/call to
- log topic
- etc.

The omission scenario this issue describes is more concretely: seaport address appearing exclusively in any of:
- tx input
- log data
- internal tx/call input
- tx output

So it is perhaps not very likely that this occurs without explicit appearances occurring in the same transaction. 

In summary: The index at present omits rare addresses (seaport) when they are only in implicit locations. 

Exclusion scenario: Tx sends calldata that contains the seaport address, but seaport contract is never called. Perhaps the address is added to some storage slot as a whitelist. 

Inclusion scenario: User interacts with seaport

Therefore, it seems reasonable to leave the UnchainedIndex as is.

**Comment by tjayrush** [comment]
> > Would it be possible for you to find the exact deployment transaction for this particular address?
> 
> 0xa7f75a8b23c0f3150b7116aae930101f42996cb35e3e596a646156926350933a

When I do

```
chifra traces --chain mainnet 0xa7f75a8b23c0f3150b7116aae930101f42996cb35e3e596a646156926350933a --fmt json | grep address
```

I get:

```
"address": "0x00000000000000adc04c56bf30ac9d3c0aaf14dc"
```

Which is the deployment. What do you get?

When I then do

```
chifra list --chain mainnet 0x00000000000000adc04c56bf30ac9d3c0aaf14dc --count
```

I get

```
0x00000000000000adc04c56bf30ac9d3c0aaf14dc	943247	7545984
```

What do you get?

When I then do:

```
chifra export --chain mainnet 0x00000000000000adc04c56bf30ac9d3c0aaf14dc --fmt json
```

It scans for quite some time and then produces a ton of data.

**Comment by tjayrush** [comment]
A few more questions:

1. Are you using the command line or API?
2. Are you using `--first_record` and/or `--max_records`?
3. Are you using `--start_block` and/or `--last_block`?

**Comment by tjayrush** [comment]
> In summary: The index at present omits rare addresses (seaport) when they are only in implicit locations. Exclusion scenario: Tx sends calldata that contains the seaport address, but seaport contract is never called. Perhaps the address is added to some storage slot as a whitelist.

Exactly. There's other places. Brilliant Soldity coder packs data destroying the ability see `appearances` because there are no longer leading zeros. The solution is sloppy and best-effort. I don't think there's much we can do about it because these things are byte strings and inline assembly leads to a near infinite search space. We do the best we can...

I'm very open to improvements, though, so thanks so much for the convo.

**Comment by tjayrush** [comment]
I'm still not sure why you're seeing zero records. Are you seeing zero records?

**Comment by perama-v** [comment]
It was my error. I was working from (bad) theory, and did not actually test. 

I now see that contracts with explicit appearances are always included, irrespective of their leading zeros.
